### PR TITLE
logindex: a refused config push leaves the installed index untouched

### PR DESCRIPTION
**Pull request by biafra23**
## Scope, honestly, up front

**No shipped host can currently reach this.** Every host dedups before the
engine sees a config: `LogIndexWatch.parse`
(`ui/src/commonMain/kotlin/io/myotis/ui/LogIndexWatch.kt:50-63`) for
desktop/Android/iOS, and `handleBuildLogIndex`
(`app/.../CommandHandler.java:381-386`) for the daemon. So this is
FFI-contract hardening against a custom, future, or direct-C-ABI caller — not
a live incident. The commits read as though an index could be lost in
production; it could not, today.

What makes it worth fixing anyway is *why* those hosts dedup.
`LogIndexWatch.parse` says it outright:

> the engine rejects a config with a duplicate address OUTRIGHT — the whole
> network's eth_getLogs would go dark over one — so dedup must happen at every
> boundary the store passes through.

The engine's own JSON boundary was the one not doing it, and the engine is the
last boundary before the damage.

## The bug

`ElReader::set_log_index_config`'s re-apply ("replace") path took the live index
out of its slot and only *then* built the replacement:

```rust
let old = slot.take().expect("checked Some above");
```

Two error arms after that point returned `false` without putting anything back,
so the installed index vanished from memory until the next successful push or a
restart. `LogIndex::new` rejects exactly one thing — duplicate watch addresses —
and `parse_log_index_config` did not screen for it, so a config that **both**
duplicated an address **and** changed an existing address's `topic0s` took the
topic-conflict arm, failed to build, and left the slot `None`. Every
`eth_getLogs` on that node then answers as if no index existed. Nothing is lost
on disk (the path checkpoints immediately before). The doc comment already
promised the opposite: *"Returns false (and installs nothing) for an invalid
config (duplicate watch addresses)"*.

### A second, quieter variant

A duplicate **alone** (no topic change) unions cleanly, so `old` *did* go back
into the slot — but only after the tail record had been retired and the coverage
it vouches for rewound away. A refused push silently dropped this run's
unconfirmed appends. Both variants have tests, and both fail on `main`:

```
a_refused_config_leaves_a_conflicting_index_exactly_as_it_was
  left: None                right: Some(48283915)   # the index vanished
a_refused_config_keeps_the_unconfirmed_coverage_it_cannot_replace
  left: Some(48283910)      right: Some(48283915)   # coverage rewound
```

## The fix

**Validate at the door** — a `duplicate_address()` guard at the top of
`set_log_index_config`, before the locks, before the checkpoint, before anything
leaves the slot. This is the "validate before anything is taken out" option of
the two available; "put `old` back on every error path" is *not* available on
the merge arm, because `LogIndex::merge` consumes the outgoing index and returns
only a `MergeError`.

**Make the commit point explicit** — every replacement candidate is now built
while the live index is still in the slot. From `slot.take()` onwards there is
no `return` at all.

### Side effects on a rejected push: none, deliberately

The bridge clear, the tail-record retire (and its `rewind_above`), the
checkpoint write and the restart-claim update all sit *after* the guard. A
refused config costs the node nothing: same index, same coverage, same restart
claim (#458's `log_index_claim`), same tail record and bridge plan, nothing
written to disk. A partial apply is worse than either outcome, because the
caller gets a bare `false` and cannot tell which parts landed.

`parse_log_index_config` rejects duplicates too, so a host gets `false` without
queueing behind the checkpoint lock — which an import can hold for as long as
merging GBs takes. Its doc comment had drifted onto `strict_bool`; moved back.

### Why the guard is sufficient

The merge path builds from `union_with`'s output, not from the screened config,
so the guard is only sound while the union cannot manufacture a duplicate the
screen never saw. It cannot: `union_with` appends a stored entry only when no
entry of that address is present, so a second copy folds into the first — which
holds even when the *stored* list is itself duplicated. That is an argument
about a function in another module, so commit 2 asserts it instead of trusting
it.

## Review round (commit 3)

An independent no-context review found five places that were unreachable behind
the new guard *by exactly the kind of argument the guard exists to stop
trusting*:

1. The merge-failure arm installed an empty index and returned **true**, with no
   log — worse than the bug being fixed, since the caller is told it succeeded.
   And because `installed_is_empty` stamps the persist clock at the 10 s floor,
   the next appender tick overwrites the full-size checkpoint written moments
   earlier. Now a `tracing::error!`.
2. That arm fell back to an index over the **pushed** config, not the union, so
   a merge failure silently *narrowed the subscription*: live `{A (imported),
   B}` + push `{B, C}` → installs `{B, C}`, A stops being indexed, under a
   warning blaming topic restrictions that do not apply. Fallback is over the
   union now.
3. `match (effective, pushed)` needed a `_` catch-all for a pair that cannot
   occur, which would silently route a union into the topic-conflict arm. The
   three values travel as one `Option` now — unrepresentable, not merely
   unreachable, and the match is exhaustive without a catch-all.
4. The **boot path** kept the consume-then-fail shape the replace path had just
   shed: `LogIndex::new(eff).ok().and_then(|p| merge(vec![p, stored_ix]))` drops
   the snapshot's index when `new` fails, installs an empty one, returns true —
   after which the next checkpoint overwrites the file that still held the
   coverage. It refuses now, leaving the file intact.
5. `deserialize` / `deserialize_v1` build `Self` directly, bypassing `new`'s
   duplicate check, and are reachable from this function's boot path via
   `LogIndex::load`. Safe only because `fingerprint()` separates entries, so
   `[A@5, A@9]` cannot hash-match a file written from `[A@5]` — an
   unreachability argument about yet another function. They apply the check
   themselves now.

The review also caught that the first two tests pass with **either** layer of
the fix alone, so neither pinned the guard. Commit 3 adds a test that holds the
checkpoint lock and asserts the refusal answers anyway — the guard's only reason
to exist. Verified to fail with the guard disabled.

## Tests

| Test | Pins |
|---|---|
| `a_refused_config_leaves_a_conflicting_index_exactly_as_it_was` | the index-vanishes variant |
| `a_refused_config_keeps_the_unconfirmed_coverage_it_cannot_replace` | the coverage/tail-record variant |
| `a_duplicate_config_is_refused_without_taking_the_checkpoint_lock` | the guard itself |
| `a_union_never_manufactures_a_duplicate_the_pushed_config_lacked` | the invariant the guard rests on |
| `duplicate_watch_addresses_are_rejected` (extended) | `duplicate_address()` directly |
| `a_watch_list_naming_one_address_twice_is_refused_at_the_parser` | the host boundary, two spellings |

Both reader tests use the offline `ElReader` helper from #458's
`restart_claim_reader_tests`, and both assert the restart claim survives.

`cargo test --workspace` green (39 suites, exit 0). `cargo clippy` introduces no
new warnings; the pre-existing deny-level `never_loop` in `el/eth/session.rs` is
untouched by this diff and fails identically on `main`. `ffi.rs` is untouched,
so no UniFFI regeneration is needed.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: rust/myotis-net/src/el/reader.rs [hunk 1]
```diff
@@ -1639,11 +1705,30 @@ impl ElReader {
...
+                            (tag, stored_ix),
+                        ]) {
+                            Ok(m) => Some(m),
+                            Err(e) => {
+                                tracing::error!(
+                                    error = ?e,
+                                    "log-index merge refused the stored snapshot against the \
+                                     pushed config; re-indexing from scratch"
+                                );
+                                None
```
[the remark this answers is not in the document]
Comment (biafra23) [reply rust/myotis-net/src/el/reader.rs:1728]: **Accepted** — fixed in 05470f69, and you're right that this is an inconsistency *this PR introduced*: commit 3 (0f918360) changed the live path's merge-failure fallback to build over the union and left the boot path's twin arm on `LogIndex::new(config)`.

Verified each step of your reasoning before changing it:

- The narrowing is real. `union_with` (logindex.rs:107-125) seeds from the pushed list and appends stored-only addresses, so an address only the snapshot watched exists in `eff` and not in `config`. Falling back to `config` drops it.
- The consequence you describe checks out too: the fallback index is empty, so `installed_is_empty` is true, so the boot path skips `note_checkpoint_adopted` — the file that still held that address's coverage is then overwritten by the next prompt checkpoint.
- `eff` moves cleanly: the `Some` arm reads only `eff.enabled` / `eff.max_speed` / `eff.backfill_paused`, all `bool`, so the `None` arm can take it by value. Compiles with no other change.
- The runtime bits come out right without the explicit setters the merge arm needs, for the reason you noted in the top-level comment: `union_with` takes the scalars from the push (logindex.rs:126-131), so `LogIndex::new(eff)` already carries them.

The log line now says "under the unioned watch-list", matching the live path.

No test: `MergeError::{Empty, ChainMismatch, TopicConflict}` are all structurally excluded here, exactly as you say, so this arm can't be reached without fault injection — same as the live-path arm it mirrors. It's a consistency fix, not a coverage one, and I'd rather say that plainly than add a test that only exercises a hand-built fallback.

**Comment by biafra23** [comment]
Thanks — replying to the two non-blocking notes from the review summary here, since a review body has no reply endpoint. The one finding is answered inline on its own thread (fixed in 05470f69).

> With the parser-level screen in `host.rs`, a duplicate pushed over the JSON FFI returns `false` **without** the reader's new `tracing::warn` ever firing (parse returns `None` first). [...] worth knowing when debugging a bare `false`

**Accepted**, and it turned out to be broader than duplicates. Checking `set_log_index_config_json` (host.rs:3073-3079), *both* parse refusals were silent — the `serde_json::from_str` failure and the `parse_log_index_config` failure — so a caller debugging a bare `false` had nothing in the engine log for any malformed config, not just this class. Fixed in 619bf7c3 by logging at the boundary that decides, which keeps `parse_log_index_config` pure (it is documented as such and is unit-tested as a pure function, so I did not want to put the log inside it).

> `union_plan` building two identical empty indexes (`pushed` + `fallback`) is the honest cost of `merge` consuming its input while `LogIndex` isn't `Clone`; the comment says as much — fine as is.

**Noted, no change.** Agreed on the diagnosis. For the record on the alternative: making `LogIndex: Clone` would be the wrong trade here — it would make cloning a multi-hundred-MB index a silent one-token operation on a type whose whole point is that it is expensive and singular, to save constructing a struct that is two empty maps and a `Vec`.

One correction to my own PR description while I am here, since it bears on how to read this: **no shipped host can reach the original bug.** `LogIndexWatch.parse` (`ui/.../LogIndexWatch.kt:50-63`) dedups for desktop/Android/iOS and `handleBuildLogIndex` (`app/.../CommandHandler.java:381-386`) dedups for the daemon. This is FFI-contract hardening against a custom or direct-C-ABI caller, not a live incident. It is at the top of the PR body, but the commit messages read as though production could lose an index, and they should not.

File: rust/myotis-engine/src/host.rs [hunk 2]
```diff
@@ -3071,10 +3071,19 @@ mod tests {
...
+    // caller either. Say it once, here, at the boundary that decides.
     let Ok(v) = serde_json::from_str::<serde_json::Value>(config_json) else {
+        tracing::warn!("log-index config is not valid JSON; ignoring the push");
         return false;
     };
     let Some(config) = parse_log_index_config(&v) else {
+        tracing::warn!(
+            "log-index config refused: a field has the wrong type, or the watch list names \
+             one address twice; ignoring the push"
+        );
```
[the remark this answers is not in the document]
Comment (biafra23) [reply rust/myotis-engine/src/host.rs:3087]: **Accepted** — fixed in 67cf6aff.

Confirmed against `parse_log_index_config`: besides a wrong scalar type it also returns `None` when a watch entry has no `address` or no `fromBlock` (both are `?` on the lookup), and when `parse_address` / `parse_word32` refuse the hex. The old warning named neither, so a host reading the log would have gone looking for the wrong mistake.

The warning now reads: a watch entry is missing `address` or `fromBlock`, a field has the wrong type or is not valid hex, or the watch list names one address twice. The function's own doc comment had the same gap and was rewritten to match.

File: rust/myotis-net/src/el/logindex.rs [hunk 3]
```diff
@@ -131,6 +131,21 @@ impl LogIndexConfig {
         })
     }
 
+    /// The duplicate watch address this config carries, if any — the ONE
+    /// reason [`LogIndex::new`] refuses a config.
+    ///
+    /// Exposed separately so a caller can refuse a bad config before it costs
+    /// anything. `LogIndex::new` is the last step of a replace, by which point
+    /// the index being replaced is already gone; a caller that must not lose
+    /// it asks here first (see `ElReader::set_log_index_config`).
```
[the remark this answers is not in the document]
Comment (biafra23) [reply rust/myotis-net/src/el/logindex.rs]: **Accepted** — fixed in 67cf6aff.

You are right that the rationale is stale, and this PR is what made it stale. `set_log_index_config` now screens the pushed config *before* the locks and before anything leaves the slot, and `reader.rs` builds every replacement candidate while the live index is still installed — so "the index being replaced is already gone" no longer describes any path.

Rewritten to state the current purpose: refuse a bad config at the door, before any lock is taken and before the live index is touched, because `LogIndex::new` sits downstream of steps that cannot be undone (`merge` consumes the outgoing index outright). The doc now also notes the second caller class you flagged separately — the deserializers ask because they build `Self` directly, bypassing `new`.

File: rust/myotis-net/src/el/logindex.rs [hunk 4]
```diff
@@ -895,7 +908,16 @@ impl LogIndex {
                     let j = parsed.watch.iter().position(|f| f.address == w.address)?;
                     coverage.push(parsed.coverage[j]);
                 }
-                Some(Self { config: config.clone(), coverage, logs: parsed.logs, cursor: parsed.cursor })
+                // Same bar as `new`: a duplicate address desynchronizes
+                // storage from coverage, and this path builds `Self` directly.
+                // Reachable with a hand-written config, since the fingerprint
+                // the file is keyed by is computed from that same config.
+                config.duplicate_address().is_none().then(|| Self {
```
[the remark this answers is not in the document]
Comment (biafra23) [reply rust/myotis-net/src/el/logindex.rs:918]: **Accepted** — fixed in 67cf6aff.

Confirmed: `duplicate_watch_addresses_are_rejected` only covers `duplicate_address()` and `LogIndex::new`. The hand-rolled re-checks in `deserialize` (v2) and `deserialize_v1` had nothing behind them.

Both are now pinned, built the way you described. Since a file is keyed by the fingerprint of the config it is read *with*, a hand-written duplicate config carries its own key rather than the writer's — so each test re-keys valid bytes to the duplicate config's fingerprint and asserts `deserialize` returns `None`. Re-keying needs no re-checksum: the checksum covers only what follows it and the fingerprint sits directly ahead of it in both layouts.

Each test also carries a control that re-keys the same bytes to a *distinct* duplicate-free config over the same addresses and asserts it loads. Without that, a `None` from the fingerprint gate or the coverage re-pair would pass as a duplicate refusal.

Verified by mutation: with either gate deleted, exactly the two new tests fail.

**Comment by biafra23** [comment]
Copilot's review overview lists three findings and no separate points of its own; each is answered on its own thread. Summarising the disposition here for the record:

1. Clarify parser warning for missing or invalid required fields — **Accepted** (67cf6aff). The warning and the parser's doc now name all of it: a watch entry missing `address` or `fromBlock`, a wrong type, hex that does not parse, and a duplicated address.
2. Update stale `duplicate_address` documentation — **Accepted** (67cf6aff). This PR is what made the old rationale stale; rewritten to the current one, and it now mentions the deserializers as callers too.
3. Add deserializer regression tests for duplicate watch addresses — **Accepted** (67cf6aff). Both the v2 and legacy-v1 direct-`Self` paths are pinned, each with a control that rules out the fingerprint gate or the coverage re-pair standing in for the duplicate check. Verified by mutation.

All checks are green.
